### PR TITLE
[dynamo][automatic_dynamic] Trigger dynamism on stride changes

### PR DESCRIPTION
**Pull request by anijain2305**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #130232

Fixes https://github.com/pytorch/pytorch/issues/129798

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @chauhang @amjames

### CODE REVIEW

File: torch/fx/experimental/symbolic_shapes.py [hunk 1]
```diff
@@ -1230,7 +1230,9 @@ class StatelessSymbolicContext(SymbolicContext):
     This will cause fresh symbols to be allocated
     """
     dynamic_sizes: DimList[DimDynamic]
+    dynamic_strides: DimList[DimDynamic]
     constraint_sizes: DimList[DimConstraint] = None
+    constraint_strides: DimList[DimConstraint] = None
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:1235]: It's hard to imagine a situation where people will actually want to put constraints on strides, but the symmetry is probably fine

File: torch/_subclasses/meta_utils.py [hunk 2]
```diff
@@ -892,15 +892,19 @@ def all_dynamic_symbolic_context(
                     )
                 t_symbolic_context = SubclassSymbolicContext(
                     dynamic_sizes=t_dynamic_sizes,
+                    dynamic_strides=t_dynamic_sizes,
                     constraint_sizes=[None] * t.ndim,
+                    constraint_strides=[None] * t.ndim,
```
Comment (ezyang) [inline torch/_subclasses/meta_utils.py:897]: It feels like a pretty strong case that this field should be defaulted

File: torch/_subclasses/meta_utils.py [hunk 3]
```diff
@@ -892,15 +892,19 @@ def all_dynamic_symbolic_context(
                     )
                 t_symbolic_context = SubclassSymbolicContext(
                     dynamic_sizes=t_dynamic_sizes,
+                    dynamic_strides=t_dynamic_sizes,
```
Comment (ezyang) [inline torch/_subclasses/meta_utils.py:895]: Are you sure??

File: torch/_subclasses/meta_utils.py [hunk 4]
```diff
@@ -892,15 +892,19 @@ def all_dynamic_symbolic_context(
...
                     constraint_sizes=[None] * t.ndim,
+                    constraint_strides=[None] * t.ndim,
                     inner_contexts=inner_contexts,  # type: ignore[arg-type]
                     tensor_source=source,
                     view_base_context=view_base_context,
                 )
             else:
                 t_symbolic_context = StatelessSymbolicContext(
                     dynamic_sizes=t_dynamic_sizes,
+                    dynamic_strides=t_dynamic_sizes,
```
Comment (ezyang) [inline torch/_subclasses/meta_utils.py:905]: ditto here, are you sure?

File: torch/fx/experimental/symbolic_shapes.py [hunk 5]
```diff
@@ -3213,8 +3230,8 @@ def _nested_int_aware_sort(tup):
                 stride[i] = self.create_symbol(
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:3230]: There's something a bit funny going on here.

Naively, if I have something like: `size = (2, 3), stride = (3, 1)` and I specify `size = (STATIC, STATIC), stride = (DYNAMIC, STATIC)`, I would expect to end up with symbolic variable allocation `size = (s0, s1), stride = (s2, 1)`; notably, s2 != s1 and I allow for `stride[0]` to vary even though everything else is constant.

However, per the logic here, the DYNAMIC entry here would actually not do anything. This is because above we would find that the stride 3 it will be present in candidates, and can be directly allocated as s1, bypassing the usage of `dynamic_dim` entirely.

This seems pretty counterintuitive to me!

It seems worth getting this right here, because it affects how downstream callers to strides go. Specifically, I feel we need a new designation like INFER, which says that you should preferentially try to infer the stride from sizes, before allocating a new stride (it's not clear to me if you should then decay to DUCK or not, I guess follow whatever we used to do.)

Comment (anijain2305) [reply torch/fx/experimental/symbolic_shapes.py:3230]: For the example, why would you want the size = (s0, s1). I agree with DYNAMIC being ignored because of candidates, but dont understand why sizes should be dynamic too.

Comment (anijain2305) [reply torch/fx/experimental/symbolic_shapes.py:3230]: I am going to send a commit which does following thing 
~~~

x = torch.arange(18).reshape((3, 6))
print(x)
print(x.shape, x.stride())
print("-----")

x1 = x[:,:3]
print(x1)
print(x1.shape, x1.stride())
print("-----")

x2 = torch.arange(9).reshape((3, 3))
print(x2)
print(x2.shape, x2.stride())
print("-----")

shape_env = ShapeEnv()
# t1 = torch.ones(2, 2, 768)
# t1 = torch.ones(2, 2, 768)
with FakeTensorMode(shape_env=shape_env) as fake_mode:
    t = fake_mode.from_tensor(
        x2,
        symbolic_context=StatelessSymbolicContext(
            dynamic_sizes=[
                DimDynamic.STATIC,
                DimDynamic.STATIC,
            ],
            dynamic_strides=[
                DimDynamic.DYNAMIC,
                DimDynamic.STATIC,
            ],
        ),
    )

    print(t, t.shape, t.stride())
~~~


Is this what you are hoping for?
~~~
FakeTensor(..., size=(3, 3), dtype=torch.int64) torch.Size([3, 3]) (s1, 1)
~~~

Comment (ezyang) [reply torch/fx/experimental/symbolic_shapes.py:3233]: I typoed, should have been static sizes lol

File: test/dynamo/test_export.py [hunk 6]
```diff
@@ -3420,6 +3420,8 @@ def test_symbool_guards(
                     x,
                     symbolic_context=StatelessSymbolicContext(
                         dynamic_sizes=[DimDynamic.DYNAMIC for _ in range(x.dim())],
+                        # stride being STATIC means infer from size
+                        dynamic_strides=[DimDynamic.STATIC for _ in range(x.dim())],
```
Comment (ezyang) [inline test/dynamo/test_export.py:3424]: INFER_STRIDE now?

Comment (ezyang) [reply test/dynamo/test_export.py:3424]: Or better yet, setup some defaulting logic

**Comment by anijain2305** [comment]
@ezyang This is ready for another round of review.

File: torch/fx/experimental/symbolic_shapes.py [hunk 7]
```diff
@@ -3210,11 +3235,15 @@ def _nested_int_aware_sort(tup):
                         if stride[i] is None
                     ], key=_nested_int_aware_sort
                 )
+                # Set INFER_STRIDE to STATIC or DUCK depending on sizes
+                dyn_stride = dynamic_strides[i]
+                if dynamic_strides[i] == DimDynamic.INFER_STRIDE:
+                    dyn_stride = DimDynamic.STATIC if are_sizes_static else DimDynamic.DUCK
```
Comment (ezyang) [inline torch/fx/experimental/symbolic_shapes.py:3241]: fwiw, DUCK here is probably wrong (this is a pre-existing condition though)

File: torch/_dynamo/variables/builder.py [hunk 8]
```diff
@@ -2157,7 +2162,9 @@ def _automatic_dynamic(
     if static_shapes:
         return StatefulSymbolicContext(
             dynamic_sizes=[DimDynamic.STATIC] * e.dim(),
+            dynamic_strides=[DimDynamic.INFER_STRIDE] * e.dim(),
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:2165]: fwiw, if sizes is static, then strides is guaranteed to be static right?

File: torch/_dynamo/variables/builder.py [hunk 9]
```diff
@@ -2211,8 +2225,36 @@ def _automatic_dynamic(
...
+                # be (10, s0) and stride can be either (s0, 1) or (s1, 1). In most cases, (s0, 1) is preferred because
+                # users are not changing both size and stride.
+                #
+                # Case 2: But for another case, lets suppose the size remains same between the two invocations but stride
+                # change. In this case, we definitely want to mark the changing stride to be DYNAMIC.
+
+                # Here, we use a hueristic to simplify determination of dynamic stride. For case 1, we will always
+                # assume that stride will be inferred (INFER_STRIDE). This might be suboptimal, where user is doing something
+                # arbitrary size and stride resizing, and we fail to trigger dynamism, but we have not seen any cases
+                # yet. For case 2, we will mark the changing dimensions DYNAMIC.
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:2245]: It's not clear to me that this needs to be fixed in this PR, but I think the determination algorithm here would be simplified if instead of modeling it as a incremental update process, we say the algorithm takes a list of tensor size+strides, and then determines what should be static/dynamic.

In this model, I think the algorithm should be something like:

1. Assess if sizes are static/dynamic, by comparing size across all entries
2. Compute the inferred strides for all of the tensors individually. If the ordering is ambiguous, consider all of these as possible. If any stride can't be inferred, that forces that stride to be dynamic across all dimensions.
3. Intersect the possible inferred strides for all tensors in the list. If this intersection is non-empty, consistently pick one order as tiebreak and use that for everything. Otherwise, if there is unreconciliable disagreement, kick out those strides as fully dynamic and then try again. This is an overly complicated description, but the basically the idea is to try to define some sort of lattice of symbol allocations and then pick the principal type

This is kind of irritatingly complicated to do, because the stride inference algorithm currently lives in symbolic_shapes.py, and we'd have to redo all the APIs. So once I understand your heuristic, I might just say that it's OK.

File: torch/_dynamo/variables/builder.py [hunk 10]
```diff
@@ -2211,8 +2225,36 @@ def _automatic_dynamic(
                             e.size(i),
                             dim,
                         )
+                        has_size_changed = True
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:2228]: This looks wrong. If size has changed in a previous iteration, you need this boolean to still be True (in this case, dim is None).

File: torch/_dynamo/variables/builder.py (hunk 9 above)
Comment (ezyang) [reply torch/_dynamo/variables/builder.py:2245]: So, it sounds like your heuristic is something like, if ALL sizes of the tensor are inferred to be static, then check to see if strides have varied. But if ANY size in the tensor is dynamic, don't bother trying to compute automatic dynamic on stride.

This seems... suspicious. Could you walk me through, step-by-step, how your heuristic computes on the repro test you have here?

**Comment by ezyang** [comment]
I feel that you have gotten in some trouble because there is a layering violation between symbolic_shapes.py and automatic_dynamic. It's hard to know if the stride is dynamic or inferred without doing the inference calculation, but that inference calculation is stuck somewhere else.

I am hoping that there is a simple heuristic we can do to start, but I don't understand the current proposed heuristic. In particular, I'm concerned about the fact that the code as written changes behavior in a path-dependent way, depending on if a size changed on this /particular/ iteration, strides are respected or not. I would prefer some sort of obviously path-independent heuristic, where no matter what order you see size-stride pairs, you always end up with the same frame state at the end. But I'm not sure how you get there without moving the stride inference logic into automatic_dynamic.

File: torch/_dynamo/variables/builder.py (hunk 9 above)
Comment (anijain2305) [reply torch/_dynamo/variables/builder.py:2245]: The heuristic is kind of simple and dumb.

First - what is INFER_STRIDE. The way I have implemented it means - let symbolic_shape infra decide if its actually static (because all size dims are static) or dynamic/duck (reuse size symbols whenever possible, or create a new symbol when tensor is strided in an unusual manner).

So, automatic_dynamic job here is to mark the stride either INFER_STRIDE or DYNAMIC.

1) If all the sizes are static, mark stride INFER_STRIDE. Let symbolic shapes decide what to do here.
2) If any of the size changes, still mark stride INFER_STRIDE. Rationale is - if size is changing, we will mark size dim DYNAMIC and symbolic_shapes will then use then reuse size symbols for the strides. Changing size usually changes strides and in predictable manner (contiguous tensors).
3) If size remains same, but the strides change (this is whats happening in generate). Mark the strides DYNAMIC.


So, lets go to the example

~~~

        @torch.compile(backend=cnt)
        def fn(x, y):
            return x * y

        for i in range(1, 4):
            x = torch.randn(4, i)

            # create a view for i > 1
            if i == 1:
                x1 = x
            else:
                x1 = x[:, 0:1]

            y = torch.randn(4, 1)
            print(x1.shape, y.shape)
            fn(x1, y)
~~~


The shape of the tensor is always same `(4, 1)` but the stride keeps changing because of the indexing operation. In this case we fall in the 3rd category, only marking the stride DYNAMIC.

**Comment by anijain2305** [comment]
@ezyang Ready now (unless CI cries)

File: torch/_dynamo/variables/builder.py [hunk 11]
```diff
@@ -2237,45 +2244,87 @@ def _automatic_dynamic(
...
+                                "automatic dynamic %s size(%s) %s != %s",
+                                name,
+                                i,
+                                e.size(i),
+                                dim,
+                            )
+                            has_size_changed = True
+                            frame_state_entry.size[i] = None
+                        elif dim is None:
+                            has_size_changed = True
```
Comment (ezyang) [inline torch/_dynamo/variables/builder.py:2296]: This is sort of overly complicated, because `has_size_changed = frame_entry.size[i] is None` after the update ;)

**Review by ezyang — APPROVED** [review]

File: torch/fx/experimental/symbolic_shapes.py [hunk 12]
```diff
@@ -3105,24 +3117,36 @@ def _create_symbolic_sizes_strides_storage_offset(
                 else:
                     r = DimDynamic.DUCK
                 dynamic_dims.append(r)
+                dynamic_strides.append(r)
             dynamic_dims = [DimDynamic.DUCK] * dim
+            dynamic_strides = [DimDynamic.INFER_STRIDE] * dim
```
Comment (retonym) [inline torch/fx/experimental/symbolic_shapes.py:3122]: Hi,

I noticed some confusing code logic between lines 3113 and 3120. It seems that the assignments made in these lines are overridden by the dynamic status being reset to DUCK on lines 3121/3122. Is there a specific reason for keeping the code this way?

Thank you.
